SalesforceFoundation/EDA
# [EDA] - Contact has Multiple Primary Affiliations to the Same Record Type

# Critical Changes

# Changes
## Update to Prevent Multiple Primary Affiliations in Certain Cases
We've fixed an issue that allowed the creation of multiple Primary Affiliations on a Contact in cases where they should've been prevented. Now, you'll see the following behavior to ensure that only one Primary Organization of the same type is allowed for Primary Affiliations for a given Contact:
- When you update the organization for a Contact’s Primary Affiliation to be the same type of organization used by a different Primary Affiliation for that Contact, all the other primary affiliations of that type will be made non-primary on that Contact.
- When you try to update the organization for a Primary Affiliation to be the same organization that is used by a different Primary Affiliation for that Contact, you’ll receive an error message and the Affiliation won’t be updated.

# Issues Closed
Closes #1097 

# New Metadata
Custom Label: AffiliationWithSameAccountExists
# Deleted Metadata

# Testing Notes

Note: Ensure Primary lookup fields are added to the contact page layout and field level security for these fields are set to visibility.

Test Scenario 1:

1. Create a Business Account B1
2. Create 2 Educational Institution Accounts E1, E2
3. Create a Contact
4. From the Contact’s related list create a Primary Affiliation to B1
5. From the Contact’s related list create a Primary Affiliation to E1
6. Confirm both the above Primary Affiliations are created with Types ‘Educational Institution’ and ‘Business Organization’
7. Confirm ‘Primary Business Organization’ and ‘Primary Educational Institution’ are populated with B1 and E1 respectively
8. Edit the Affiliation created on step 4 → Change Organization field from B1 to E2
9. Confirm ‘Primary Business Organization’ on Contact is cleared (as Expected)
10. Confirm ‘Primary Educational Institution’ on the Contact is updated to E2 (as expected)
11. *Bug*: Now there are 2 affiliation of same type ‘Educational Institution’ as expected but the 2 affiliations are now Primary. Expected: Only the latest Affiliation of same type needs to be Primary. In this case, only E2 needs to be Primary and E1 needs to be updated to non-primary.
12. *Fix*: After the fix, only E2 needs to be Primary and E1 needs to be updated to non-primary.



Test Scenario 2:

1. Create 1 Business Account B1
2. Create 1 Educational Institution Account E1
3. Create a Contact
4. From the Contact’s related list create a Primary Affiliation to B1
5. From the Contact’s related list create a Primary Affiliation to E1
6. Confirm both the above Primary Affiliations are created with Types ‘Educational Institution’ and ‘Business Organization’
7. Confirm ‘Primary Business Organization’ and ‘Primary Educational Institution’ are populated with B1 and E1
8. Edit the Affiliation created on step 4 → Change Organization field from B1 to E1
9. Confirm ‘Primary Business Organization’ on Contact is cleared (as Expected)
10. Confirm ‘Primary Educational Institution’ on the Contact is still to E1 (as Expected)
11. *Bug*: Now there are 2 affiliation of same type ‘Educational Institution’ as expected but the 2 affiliations are now Primary. 
12. *Fix*: When changing a Primary Affiliation’s Account to another Account, if there is already another Primary affiliation to the same Account then stop the update and show an error to the user. Something like, “Update stopped. There is another primary Affiliation on this Contact with the same Account.”
